src/core/lib/support/time.c

```diff
@@ -80,103 +80,62 @@ gpr_timespec gpr_inf_past(gpr_clock_type type) {
   return out;
 }
 
-/* TODO(ctiller): consider merging _nanos, _micros, _millis into a single
-   function for maintainability. Similarly for _seconds, _minutes, and _hours */
-
-gpr_timespec gpr_time_from_nanos(int64_t ns, gpr_clock_type type) {
-  gpr_timespec result;
-  result.clock_type = type;
-  if (ns == INT64_MAX) {
-    result = gpr_inf_future(type);
-  } else if (ns == INT64_MIN) {
-    result = gpr_inf_past(type);
-  } else if (ns >= 0) {
-    result.tv_sec = ns / GPR_NS_PER_SEC;
-    result.tv_nsec = (int32_t)(ns - result.tv_sec * GPR_NS_PER_SEC);
+static gpr_timespec get_timespec_by_different_second_unit(int64_t t, int64_t n,
+                                                          gpr_clock_type type) {
+  gpr_timespec out;
+  if (t == INT64_MAX) {
+    out = gpr_inf_future(type);
+  } else if (t == INT64_MIN) {
+    out = gpr_inf_past(type);
   } else {
-    /* Calculation carefully formulated to avoid any possible under/overflow. */
-    result.tv_sec = (-(999999999 - (ns + GPR_NS_PER_SEC)) / GPR_NS_PER_SEC) - 1;
-    result.tv_nsec = (int32_t)(ns - result.tv_sec * GPR_NS_PER_SEC);
+    if (t >= 0) {
+      out.tv_sec = t / n;
+    } else {
+      out.tv_sec = (-((n - 1) - (t + n)) / n) - 1;
+    }
+    out.tv_nsec = (int32_t)((t - out.tv_sec * n) * GPR_NS_PER_SEC / n);
+    out.clock_type = type;
   }
-  return result;
+  return out;
 }
 
-gpr_timespec gpr_time_from_micros(int64_t us, gpr_clock_type type) {
-  gpr_timespec result;
-  result.clock_type = type;
-  if (us == INT64_MAX) {
-    result = gpr_inf_future(type);
-  } else if (us == INT64_MIN) {
-    result = gpr_inf_past(type);
-  } else if (us >= 0) {
-    result.tv_sec = us / 1000000;
-    result.tv_nsec = (int32_t)((us - result.tv_sec * 1000000) * 1000);
+static gpr_timespec get_timespec_by_hour_or_minute_unit(int64_t t, int64_t n,
```

Comment:
Following the previous comment, I suggest changing the name to `to_seconds_from_above_second_time`. This follows the same naming schema and doesn't limit you to only "hour or minute". This function would be used to calculate seconds in a week, or in a year. The current name misrepresents what it's able to do. 

Same comment regarding `t` and `n`: how about renaming `n` to `secs_per_unit` (note this is different from the previous suggestion for `n`) and `t` to `time_in_units`?